phoenix-core/src/main/java/org/apache/phoenix/optimize/QueryOptimizer.java

```diff
@@ -91,8 +91,23 @@ public QueryPlan optimize(PhoenixStatement statement, SelectStatement select, Co
     }
     
     public QueryPlan optimize(QueryPlan dataPlan, PhoenixStatement statement, List<? extends PDatum> targetColumns, ParallelIteratorFactory parallelIteratorFactory) throws SQLException {
-        List<QueryPlan>plans = getApplicablePlans(dataPlan, statement, targetColumns, parallelIteratorFactory, true);
-        return plans.get(0);
+        List<QueryPlan> plans = getApplicablePlans(dataPlan, statement, targetColumns, parallelIteratorFactory, false);
+        if (plans.size() == 1) {
+            return plans.get(0);
+        }
+
+        // Get the best plan based on their costs. Costs will be ZERO if stats are not
+        // available, thus the first plan will be returned.
+        Cost minCost = null;
+        QueryPlan bestPlan = null;
+        for (QueryPlan plan : plans) {
+            Cost cost = plan.getCost();
+            if (minCost == null || cost.compareTo(minCost) < 0) {
+                minCost = cost;
+                bestPlan = plan;
+            }
+        }
+        return bestPlan;
```

Comment:
This will ignore any index hints used in the query.

Reply:
Thanks for the feedback, @katameru! We need to define a strategy how hints work in the cost-based optimizer mode. Your input would be very much appreciated!
Since this is the first issue in which I've brought "cost-based" in, the idea is to plug-in the framework and pieces of code here and there needed to make it work in general. We can see there's still a lot to do to refine this framework, and as I dig further into other issues under PHOENIX-4313. Right now I think adding a option to disable cost-based optimizer (pls see my latest ci) should be enough.

Reply:
I'd argue that the observable behavior should stay the same, so a hinted query plan should be used above all others. Currently we could just sidestep the cost calculations in favor of a hinted plan, but I think that having a way to influence the optimizer via hints could be very useful.

Reply:
I agree with @katameru. If you plug in your optimization code to impact the sort order done by QueryOptimizer.orderPlansBestToWorst() when getApplicablePlans() is called, we'd keep the behavior that the hints would override the optimizer (whether cost-based is enabled or not).

I still think we need more tests. We've found that if the tests don't get added with the initial commit, they unfortunately never get added.

Reply:
Thank you both for the input. I've change the logic for hinted plans as well as added new tests. Please review again.

Reply:
QueryOptimize.optimize() is not the only entry point to the optimizer. For example, DeleteCompiler calls both getApplicablePlans() and getBestPlan(). I think it'd be best if you hid this logic at the end of the private getApplicablePlans since all calls channel through that. That way all optimizer decisions will use cost if it's enabled. You could conditional order the plans based on their cost if cost-based is enabled here:

            return hintedPlan != null ? plans : isCostBasedEnabled ? orderPlansBasedOnCost(plans) : orderPlansBestToWorst(select, plans, stopAtBestPlan);

Reply:
I'm still not seeing this change. The only existing method that should change in QueryOptimizer is this private one:

    private List<QueryPlan> getApplicablePlans(QueryPlan dataPlan, PhoenixStatement statement, List<? extends PDatum> targetColumns, ParallelIteratorFactory parallelIteratorFactory, boolean stopAtBestPlan) throws SQLException {

The only change necessary should be to call a new method that orders the plans according to cost.